### PR TITLE
fix(cdn): compute cdn_items_sync metrics

### PR DESCRIPTION
**Pull request by yesnault**
Signed-off-by: Yvonnick Esnault <yvonnick.esnault@corp.ovh.com>

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#15326.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package and Analyze
  * Docker ✔ 
  * Sonar ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Review by fsamin — APPROVED** [review]

**Review by sguiheux — APPROVED** [review]

File: engine/cdn/cdn_metrics.go [hunk 1]
```diff
@@ -157,7 +160,7 @@ func (s *Service) ComputeMetrics(ctx context.Context) {
 				telemetry.Record(ctxItem, s.Metrics.ItemToDelete, stat.Number)
 			}
 
-			storageStats, err = storage.CountItemUnitToDelete(s.mustDBWithCtx(ctx))
+			storageStats, err = storage.CountItemUnitToDelete(s.mustDBWithCtx(ctx)) // 93.706 ms
```
Comment (richardlt) [inline engine/cdn/cdn_metrics.go:163]:
```suggestion
			storageStats, err = storage.CountItemUnitToDelete(s.mustDBWithCtx(ctx))
```

File: engine/cdn/cdn_metrics.go [hunk 2]
```diff
@@ -107,46 +110,46 @@ func (s *Service) ComputeMetrics(ctx context.Context) {
...
+						if previous, ok := s.storageUnitLags.Load(key); ok {
+							s.storageUnitPreviousLags.Store(key, previous)
+						}
+						s.storageUnitLags.Store(key, lag)
 						break
 					}
 				}
 			}
-			itemsToDelete, err := item.CountItemsToDelete(s.mustDBWithCtx(ctx))
+			itemsToDelete, err := item.CountItemsToDelete(s.mustDBWithCtx(ctx)) // 27.004 ms
```
Comment (richardlt) [inline engine/cdn/cdn_metrics.go:152]:
```suggestion
			itemsToDelete, err := item.CountItemsToDelete(s.mustDBWithCtx(ctx))
```

File: engine/cdn/cdn_metrics.go [hunk 3]
```diff
@@ -184,3 +194,22 @@ func (s *Service) ComputeMetrics(ctx context.Context) {
 		}
 	}
 }
+
+func (s *Service) countItemsForUnit(ctx context.Context, storageUnit storage.Interface) []storage.Stat {
+	types := []sdk.CDNItemType{sdk.CDNTypeItemStepLog, sdk.CDNTypeItemServiceLog, sdk.CDNTypeItemArtifact}
+	var storageStats []storage.Stat
+	for _, typ := range types {
+		suStats, err := storage.CountItemsForUnit(s.mustDBWithCtx(ctx), storageUnit.ID(), string(typ)) // 0.033 ms
```
Comment (richardlt) [inline engine/cdn/cdn_metrics.go:202]:
```suggestion
		suStats, err := storage.CountItemsForUnit(s.mustDBWithCtx(ctx), storageUnit.ID(), string(typ))
```

File: engine/cdn/cdn_metrics.go [hunk 4]
```diff
@@ -107,46 +110,46 @@ func (s *Service) ComputeMetrics(ctx context.Context) {
...
-				if err != nil {
-					log.Error(ctx, "cdn> Unable to compute CountItemsByUnit for %s: %v", su.Name(), err)
-					continue
-				}
-				for i := range suStats {
-					s := suStats[i]
-					s.StorageName = su.Name()
-					storageStats = append(storageStats, s)
-				}
+				storageStats = append(storageStats, s.countItemsForUnit(ctx, su)...) // 0.066 ms
```
Comment (richardlt) [inline engine/cdn/cdn_metrics.go:125]:
```suggestion
				storageStats = append(storageStats, s.countItemsForUnit(ctx, su)...)
```

File: engine/cdn/cdn_metrics.go [hunk 5]
```diff
@@ -107,46 +110,46 @@ func (s *Service) ComputeMetrics(ctx context.Context) {
...
-				if err != nil {
-					log.Error(ctx, "cdn> Unable to compute CountItemsByUnit for %s: %v", bu.Name(), err)
-					continue
-				}
-				for i := range bufferStats {
-					b := bufferStats[i]
-					b.StorageName = bu.Name()
-					storageStats = append(storageStats, b)
-				}
+				storageStatsBuffer = append(storageStatsBuffer, s.countItemsForUnit(ctx, bu)...) // 0.066 ms
```
Comment (richardlt) [inline engine/cdn/cdn_metrics.go:115]:
```suggestion
				storageStatsBuffer = append(storageStatsBuffer, s.countItemsForUnit(ctx, bu)...)
```

File: engine/cdn/cdn_metrics.go [hunk 6]
```diff
@@ -88,15 +90,16 @@ func (s *Service) initMetrics(ctx context.Context) error {
...
 	defer tickStatsItems.Stop()
 	for {
 		select {
 		case <-ctx.Done():
 			return
 		case <-tickStatsItems.C:
+			start := time.Now()
 			// All Items by type
-			allItemsByType, err := item.CountItems(s.mustDBWithCtx(ctx))
+			allItemsByType, err := item.CountItems(s.mustDBWithCtx(ctx)) // 308.852 ms
```
Comment (richardlt) [inline engine/cdn/cdn_metrics.go:102]:
```suggestion
			allItemsByType, err := item.CountItems(s.mustDBWithCtx(ctx))
```

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#15333.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package and Analyze
  * Docker ✔ 
  * Sonar ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#15333.1 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package and Analyze
  * Docker ✔ 
  * Sonar ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#15333.2 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package and Analyze
  * Docker ✔ 
  * Sonar ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Pull request merged**